### ISSUE TITLE
A bug in generate

### ISSUE THREAD

**Issue by Jiang-Stan** [lucidrains/soundstorm-pytorch#28]
https://github.com/lucidrains/soundstorm-pytorch/blob/4fe623e3602ca3998f86a4a8c08360412767e7b2/soundstorm_pytorch/soundstorm.py#L941

I think the score here should be logits.softmax, rather than 1-logits.softmax. logits in training is calculated normally, so the higher the logits, the better it should be.

I also tested both code, and found that if 1-logits is used, unmasked tokens would change in every step of first layer. But if it is replaced by logits, most of unmasked tokens of first layer predicted are kept in the next step.

**Comment by lucidrains** [comment]
@Jiang-Stan hey Jiang

I later on mask out by [highest descending score](https://github.com/lucidrains/soundstorm-pytorch/blob/main/soundstorm_pytorch/soundstorm.py#L956-L960) though. by your logic, we would be masking out the tokens the network is most confident about?

**Comment by Jiang-Stan** [comment]
Thanks for your reply!
I think you are right. Scores here should be calc by 1-softmax. Sorry for bothering you!
But I find another bug here, and I was mislead by it.
https://github.com/lucidrains/soundstorm-pytorch/blob/4fe623e3602ca3998f86a4a8c08360412767e7b2/soundstorm_pytorch/soundstorm.py#L956
argsort can return the sorted index of scores, but the mask used here should be in pair with original score. Therefore I think the following code should be added:
```
scores_sorted = scores.argsort(dim = -1, descending = True)
idx = torch.linspace(0, scores.shape[1]-1, scores.shape[1], dtype=torch.int64, device = device)
sort_of_each_score = torch.zeros(scores_sorted.shape, dtype=torch.int64, device = device)
for i in range(len(scores_sorted)):
    sort_of_each_score[i][scores_sorted] = idx

mask_num_tokens = rearrange(mask_num_tokens, 'b -> b 1')

mask = sort_of_each_score < mask_num_tokens
```

**Comment by lucidrains** [comment]
@Jiang-Stan oh my god... 🤦 thank you so much, i think you are correct

can you see if the extra chained `.argsort` fixes it for you?

**Comment by hjzzju** [comment]
why not directly scores.argsort(dim = -1, descending = False), seems equal to scores.argsort(dim=-1, descending=True).argsort(dim=-1)
